Tests: Housekeeping and Clean Sweep of Sevice/Logging suite

jakub-vavra-cz opened the pull request.

danlavu commented on src/tests/system/tests/test_logging.py:
```diff
@@ -45,38 +50,91 @@ def test_logging__default_settings_logs_debug_level(client: Client):
+    :customerscenario: False
+    """
+    client.sssd.common.local()
+    client.sssd.default_domain = "local"
```
I think this is silly, I created https://github.com/SSSD/sssd-test-framework/pull/162/files so it's a part of sssd.common.local() , added you as a reviewer.

danlavu commented on src/tests/system/tests/test_logging.py:
```diff
@@ -45,38 +50,91 @@ def test_logging__default_settings_logs_debug_level(client: Client):
+    client.sssd.domain["debug_level"] = "9"
+    client.sssd.clear(logs=True)
+    client.sssd.start(debug_level=None)
+    log_str = client.fs.read(client.sssd.logs.domain())
```
nitpick, line break between the setup and the steps.

danlavu commented on src/tests/system/tests/test_logging.py:
```diff
@@ -45,38 +50,91 @@ def test_logging__default_settings_logs_debug_level(client: Client):
+    client.sssd.default_domain = "local"
+    client.sssd.clear(logs=True)
+    client.sssd.start(debug_level=None)
+    log_str = client.fs.read(client.sssd.logs.domain())
```
nitpick, line break between the setup and steps

danlavu commented on src/tests/system/tests/test_logging.py:
```diff
@@ -100,64 +158,107 @@ def test_logging__default_settings_logs_domain_configuration_errors(client: Clie
+    client.sssd.enable_responder("ifp")
+    client.sssd.start(debug_level=None, raise_on_error=False)
+
+    time.sleep(60)
```
Can this be lowered? ldap_network_timeout is 1, it looks like it can be.

jakub-vavra-cz replied:
I tried 30 but it was flaky more failing than passing.

alexey-tikhonov replied:
How exactly did it fail?

jakub-vavra-cz replied:
The offline messages were not written in logs.
Switched to polling loops instead of fixed sleeps.

danlavu commented on src/tests/system/tests/test_logging.py:
```diff
@@ -100,64 +158,107 @@ def test_logging__default_settings_logs_domain_configuration_errors(client: Clie
+    client.sssd.domain["ldap_opt_timeout"] = "1"
+    client.sssd.domain["ldap_network_timeout"] = "1"
+    client.sssd.start(debug_level=None, raise_on_error=False)
+    time.sleep(95)
```
95 seconds! I know overhead and time are important to you, but why does this value have to be so high?

alexey-tikhonov replied:
(1) Can this test be merged with previous test?
(2) How exactly does it fail with low timeout?

jakub-vavra-cz replied:
Merged them to together and added polling loops instead of fixed sleeps.

danlavu reviewed (commented):
Some questions, nitpicks, but it's great, thank you!

alexey-tikhonov commented on src/tests/system/tests/test_logging.py:
```diff
@@ -45,38 +50,91 @@ def test_logging__default_settings_logs_debug_level(client: Client):
     client.fs.copy("/var/log/sssd", "/tmp/copy")
-    assert client.auth.ssh.password("user1", "Secret123"), "Login failed!"
+    assert client.auth.parametrize("ssh").password("user1", "Secret123"), "User failed login!"
     assert not client.host.conn.run("diff /var/log/sssd /tmp/copy").stdout, "Debug messages were generated!"
```
This skips files that didn't exist at start:
```    
    Only in /var/log/sssd: gpo_child.log
    Only in /var/log/sssd: krb5_child.log
    Only in /var/log/sssd: sssd_kcm.log
```

jakub-vavra-cz replied:
ack

jakub-vavra-cz commented:
> This uncovers following things that likely need to be fixed in the code:
> 
> (1) `test_logging__default_settings_does_not_log_user_logins (ipa)`
> 
> ```
>     < (2025-03-26 14:43:29): [be[test]] [ipa_subdomains_refresh_passkey_done] (0x3f7c0): [RID#1] Passkey feature is not configured on IPA server
>     < (2025-03-26 14:43:29): [be[test]] [ipa_domain_resolution_order_done] (0x3f7c0): [RID#1] Failed to get the domains' resolution order configuration from the server [22]: Invalid argument
>     < (2025-03-26 14:43:29): [be[test]] [ipa_subdomains_handler_done] (0x0020): [RID#1] Unable to refresh subdomains [22]: Invalid argument
> ```
> 
> (2) `test_logging__default_settings_does_not_log_user_logins (ldap)`
> 
> ```
>     < (2025-03-26 15:17:43): [ssh] [cache_req_common_process_dp_reply] (0x3f7c0): [CID#1] CR #0: Data Provider Error: 3, 22, Invalid argument
> ```
> 
> (3) `test_logging__default_settings_does_not_log_user_logins (samba)` (and the same with 'proxy')
> 
> ```
>     < (2025-03-26 15:43:21): [ssh] [cache_req_common_process_dp_reply] (0x3f7c0): [CID#1] CR #0: Could not get account info [1432158215]: DP target is not configured
> ```
> 
> Most probably all fixes will be decrease log level, but to be confirmed.

For ipa this is the config:

[sssd]
services = nss, pam, ssh
domains = test

[domain/test]
ipa_domain = ipa.test
id_provider = ipa
access_provider = ipa
ipa_server = master.ipa.test
dyndns_update = False
fallback_homedir = /home/%%u

danlavu reviewed (commented):
Last bit of nitpicks, we should rename the tests. I think we should exclude the 'default_' and assume that is the default and state the differences. 

Linebreaks, let's add a linebreak to distinguish the start of a new step.

jakub-vavra-cz commented on src/tests/system/tests/test_logging.py:
```diff
@@ -45,38 +50,93 @@ def test_logging__default_settings_logs_debug_level(client: Client):
 @pytest.mark.importance("low")
 @pytest.mark.topology(KnownTopology.Client)
-def test_logging__default_settings_does_not_log_user_logins(client: Client):
+def test_logging__custom_settings_logs_debug_level(client: Client):
```
done

jakub-vavra-cz commented on src/tests/system/tests/test_logging.py:
```diff
@@ -45,38 +50,93 @@ def test_logging__default_settings_logs_debug_level(client: Client):
+    log_str = client.fs.read(client.sssd.logs.domain())
+    sssctl_set = client.host.conn.run("sssctl debug-level --domain local 9")
+    sssctl_get = client.host.conn.run("sssctl debug-level --domain local")
+    assert "level = 0x0070" in log_str, "Logs should contain debug level = 0x0070!"
```
done

jakub-vavra-cz commented on src/tests/system/tests/test_logging.py:
```diff
@@ -45,38 +50,93 @@ def test_logging__default_settings_logs_debug_level(client: Client):
+@pytest.mark.integration
+@pytest.mark.importance("low")
+@pytest.mark.topology(KnownTopologyGroup.AnyProvider)
+def test_logging__default_settings_does_not_log_user_logins(client: Client, provider: GenericProvider):
```
done

jakub-vavra-cz commented on src/tests/system/tests/test_logging.py:
```diff
@@ -100,64 +160,83 @@ def test_logging__default_settings_logs_domain_configuration_errors(client: Clie
-@pytest.mark.ticket(bz=1893159)
 @pytest.mark.topology(KnownTopology.LDAP)
-def test_logging__default_settings_logs_offline_errors(client: Client):
+def test_logging__default_settings_logs_offline_dns(client: Client):
```
test_logging_dns_resolution_issue_in_logs

danlavu replied:
It does, thank you.

jakub-vavra-cz commented on src/tests/system/tests/test_logging.py:
```diff
@@ -100,64 +160,83 @@ def test_logging__default_settings_logs_domain_configuration_errors(client: Clie
+    assert (
+        "Failed to resolve server 'typo.invalid': Domain name not found" in logs
+    ), "'Domain name not found' error message are not in logs!"
     result = client.sssctl.domain_status(client.sssd.default_domain)
```
This ugly format was done by black.

jakub-vavra-cz commented on src/tests/system/tests/test_logging.py:
```diff
@@ -100,64 +160,83 @@ def test_logging__default_settings_logs_domain_configuration_errors(client: Clie
-@pytest.mark.ticket(bz=1416150)
 @pytest.mark.topology(KnownTopology.LDAP)
-def test_logging__default_settings_logs_to_syslog_when_ldap_is_offline(client: Client):
+def test_logging__default_settings_logs_offline_errors(client: Client, ldap: LDAP):
```
Changed to test_logging_offline_errors_are_written_to_logs_and_syslog

jakub-vavra-cz commented on src/tests/system/tests/test_logging.py:
```diff
@@ -100,64 +160,83 @@ def test_logging__default_settings_logs_domain_configuration_errors(client: Clie
+    client.sssd.start(debug_level=None, raise_on_error=False)
+    if not client.sssd.default_domain:
+        client.sssd.default_domain = "ldap.test"
+    for _ in range(1, 10):
```
done

jakub-vavra-cz commented on src/tests/system/tests/test_logging.py:
```diff
@@ -100,64 +160,83 @@ def test_logging__default_settings_logs_domain_configuration_errors(client: Clie
+    assert (
+        "Connection timed out [ldap_network_timeout]" in logs
+    ), "'Connection timed out [ldap_network_timeout]' error message is not in logs!"
+    for _ in range(1, 10):
```
done

danlavu commented on src/tests/system/tests/test_logging.py:
```diff
@@ -45,38 +50,92 @@ def test_logging__default_settings_logs_debug_level(client: Client):
+    assert "level = 0x0070" in log_str, "Logs should contain debug level = 0x0070!"
+    assert sssctl_set.rc == 0, "'sssctl debug-level' command failed!"
+    assert sssctl_get.rc == 0, "'sssctl debug-level' command failed!"
+    assert " 0x2f7f0" in sssctl_get.stdout, "Sssctl debug-level output should contain 0x2f7f0"
```
space? 0x2f7f0, does just the string return more than one result?

'Sssctl' is strange since it's a command, 'sssctl'

and an exclamation at the end

jakub-vavra-cz replied:
done

danlavu reviewed (approved):
Thank you @jakub-vavra-cz , just a few nitpicks on one line. Approving, please make the small edit.

alexey-tikhonov commented:
@jakub-vavra-cz, could you please rebase?

jakub-vavra-cz commented:
> Thank you @jakub-vavra-cz , just a few nitpicks on one line. Approving, please make the small edit.

I simplified the test names, added several line breaks and edited the error message as requested.

jakub-vavra-cz commented:
> @jakub-vavra-cz, could you please rebase?

rebased & pushed.

alexey-tikhonov commented:
> (1) `test_logging__default_settings_does_not_log_user_logins (ipa)`
> 
> ```
> [be[test]] [ipa_subdomains_refresh_passkey_done] (0x3f7c0): [RID#1] Passkey feature is not configured on IPA server
> [be[test]] [ipa_domain_resolution_order_done] (0x3f7c0): [RID#1] Failed to get the domains' resolution order configuration from the server [22]: Invalid argument
> [be[test]] [ipa_subdomains_handler_done] (0x0020): [RID#1] Unable to refresh subdomains [22]: Invalid argument
> ```

@justin-stephenson , do you remember the reason why 2 messages above are logged as `SSSDBG_IMPORTANT_INFO`? (fa326be9cb29d97d6000e783b02656f60f7f8fb2)

alexey-tikhonov commented:
> (2) `test_logging__default_settings_does_not_log_user_logins (ldap)`
> 
> ```
> [ssh] [cache_req_common_process_dp_reply] (0x3f7c0): [CID#1] CR #0: Data Provider Error: 3, 22, Invalid argument
> ```
> 
> (3) `test_logging__default_settings_does_not_log_user_logins (samba)` (and the same with 'proxy')
> 
> ```
> [ssh] [cache_req_common_process_dp_reply] (0x3f7c0): [CID#1] CR #0: Could not get account info [1432158215]: DP target is not configured
> ```

@jakub-vavra-cz, could you please run those 2 tests with 'debug_level = 9' in the '[ssh]' and domain sections of sssd.conf and send me logs?

jakub-vavra-cz commented on src/tests/system/tests/test_logging.py:
```diff
@@ -100,64 +160,87 @@ def test_logging__default_settings_logs_domain_configuration_errors(client: Clie
+    :title: Default debug_level logs backend dns resolution issue
     :setup:
-        1. Configure SSSD with an invalid uri and enable ifp responder
+        1. Configure SSSD with an invalid uri and enable
```
fixed

jakub-vavra-cz commented on src/tests/system/tests/test_logging.py:
```diff
@@ -100,64 +160,87 @@ def test_logging__default_settings_logs_domain_configuration_errors(client: Clie
         2. Check default domain status
     :expectedresults:
-        1. Logs contain connection errors
+        1. Logs contain dns error message
```
done

jakub-vavra-cz commented on src/tests/system/tests/test_logging.py:
```diff
@@ -100,64 +160,87 @@ def test_logging__default_settings_logs_domain_configuration_errors(client: Clie
-@pytest.mark.ticket(bz=1893159)
 @pytest.mark.topology(KnownTopology.LDAP)
-def test_logging__default_settings_logs_offline_errors(client: Client):
+def test_logging_dns_resolution_issue_in_logs(client: Client):
```
@shridhargadekar  It did not work without it. My guess is that the responder checks for backend status early so the error message is written without long wait or search for an user.

justin-stephenson commented:
> > (1) `test_logging__default_settings_does_not_log_user_logins (ipa)`
> > ```
> > [be[test]] [ipa_subdomains_refresh_passkey_done] (0x3f7c0): [RID#1] Passkey feature is not configured on IPA server
> > [be[test]] [ipa_domain_resolution_order_done] (0x3f7c0): [RID#1] Failed to get the domains' resolution order configuration from the server [22]: Invalid argument
> > [be[test]] [ipa_subdomains_handler_done] (0x0020): [RID#1] Unable to refresh subdomains [22]: Invalid argument
> > ```
> 
> @justin-stephenson , do you remember the reason why 2 messages above are logged as `SSSDBG_IMPORTANT_INFO`? ([fa326be](https://github.com/SSSD/sssd/commit/fa326be9cb29d97d6000e783b02656f60f7f8fb2))

I'm afraid I don't remember. Would `SSSDBG_MINOR_FAILURE` be more appropriate? Feel free to adjust as needed.

alexey-tikhonov commented:
> > > (1) `test_logging__default_settings_does_not_log_user_logins (ipa)`
> > > ```
> > > [be[test]] [ipa_subdomains_refresh_passkey_done] (0x3f7c0): [RID#1] Passkey feature is not configured on IPA server
> > > [be[test]] [ipa_domain_resolution_order_done] (0x3f7c0): [RID#1] Failed to get the domains' resolution order configuration from the server [22]: Invalid argument
> > > [be[test]] [ipa_subdomains_handler_done] (0x0020): [RID#1] Unable to refresh subdomains [22]: Invalid argument
> > > ```
> > 
> > 
> > @justin-stephenson , do you remember the reason why 2 messages above are logged as `SSSDBG_IMPORTANT_INFO`? ([fa326be](https://github.com/SSSD/sssd/commit/fa326be9cb29d97d6000e783b02656f60f7f8fb2))
> 
> I'm afraid I don't remember. Would `SSSDBG_MINOR_FAILURE` be more appropriate? Feel free to adjust as needed.

#7928

alexey-tikhonov commented:
@jakub-vavra-cz, ping:

> Please rebase on the top of merged #7928

and https://github.com/SSSD/sssd/pull/7891#issuecomment-2798793240

jakub-vavra-cz commented:
> @jakub-vavra-cz, ping:
> 
> > Please rebase on the top of merged #7928
> 
> and [#7891 (comment)](https://github.com/SSSD/sssd/pull/7891#issuecomment-2798793240)
 @alexey-tikhonov  rebased onto master

alexey-tikhonov commented:
Looks like there are 3 issues:

(1) man page doesn't mention that 'ldap' is supported

(2) imo, it's wrong that ldap-provider returns 'Invalid argument' for 'localhost'

(3) wrt 'DP target is not configured' seen with a default debug level - I'm not so sure.
On the one hand, if somebody explicitly configures 'ssh' in 'services' list it's reasonable to expect that configured domain supports this service and to warn if it doesn't.
On the other hand, in theory there maybe multiple domains configured, with only some of them supporting 'hostid' lookups. In this setup it is expected that other domains will fail this lookup and warning isn't needed.

shridhargadekar commented on src/tests/system/tests/test_logging.py:
```diff
@@ -45,40 +50,94 @@ def test_logging__default_settings_logs_debug_level(client: Client):
-@pytest.mark.ticket(bz=1893159)
 @pytest.mark.topology(KnownTopology.Client)
-def test_logging__default_settings_logs_domain_configuration_errors(client: Client):
+def test_logging__domain_configuration_errors(client: Client):
```
This could be explored more. Not a change request, but a suggestion. 
For example, at least one option is misconfigured but sssd still starts. Assert the related error in log.

shridhargadekar replied:
just a suggestion.

shridhargadekar commented:
LGTM, Apart from above added comments from Alexey.
